neqo-transport/src/tparams.rs

```diff
@@ -666,49 +668,57 @@ impl TransportParametersHandler {
     }
 
     fn compatible_upgrade(&mut self, remote_tp: &TransportParameters) -> Res<()> {
-        if let Some((current, other)) = remote_tp.get_versions() {
-            qtrace!(
-                "Peer versions: {current:x} {other:x?}; config {:?}",
-                self.versions,
-            );
-
-            if self.role == Role::Client {
-                let chosen = Version::try_from(current)?;
-                if self.versions.compatible().any(|&v| v == chosen) {
-                    Ok(())
-                } else {
-                    qinfo!(
-                        "Chosen version {current:x} is not compatible with initial version {:x}",
-                        self.versions.initial().wire_version(),
-                    );
-                    Err(Error::TransportParameter)
-                }
+        if self.version_selected {
+            // A second call to this is only possible on the server,
+            // if we need to send a TLS HelloRetryRequest.
+            debug_assert_eq!(self.role, Role::Server);
+            return Ok(());
+        }
+
+        let Some((current, other)) = remote_tp.get_versions() else {
+            return Ok(());
+        };
```

Comment:
Early return here makes the changes look more significant than they are.  The change is the block above and the two places where the `version_selected` flag is set to `true`.